aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -654,6 +654,70 @@ Tensor& lgamma_out(Tensor& result, const Tensor& self) { return unary_op_impl_fl
 Tensor lgamma(const Tensor& self) { return unary_op_impl_float(self, lgamma_stub); }
 Tensor& lgamma_(Tensor& self) { return unary_op_impl_(self, at::lgamma_out); }
 
+static inline TensorIterator build_frexp_iter(const Tensor &self,
+                                              Tensor mantissa, Tensor exponent) {
+
+  TensorIteratorConfig config;
+
+  config.set_check_mem_overlap(true)
+  .add_output(mantissa)
+  .add_output(exponent);
+
+  // gpu_kernel_multiple_outputs also has issues of handling non contiguous tensors
+  // Reference: https://github.com/pytorch/pytorch/pull/51097
+  if (self.device().type() == DeviceType::CUDA) {
+    config.add_input(self.contiguous());
+  } else {
+    config.add_input(self);
+  }
+
+  return config
+    .cast_common_dtype_to_outputs(true)
+    .promote_inputs_to_common_dtype(true)
+    .promote_integer_inputs_to_float(true)
+    .enforce_safe_casting_to_output(true)
+    .build();
+}
+
+static inline void frexp_check(const Tensor& self) {
+  // torch.frexp currently does not support integral dtypes for cuda tensors
+  // due to the casting issues of gpu_kernel_multiple_outputs
+  // Reference: https://github.com/pytorch/pytorch/pull/51097
+  TORCH_CHECK(at::isFloatingType(self.scalar_type()) || self.device().is_cpu(),
+              "frexp is not implemented for ",
+              self.scalar_type(),
+              " on CUDA device.");
```

Comment:
The `gpu_kernel_multiple_outputs` uses `multi_outputs_unroll` for unrolling, and it seems `multi_outputs_unroll` cannot handling `casting` as it is implemented by `LoadWithoutCast` and `StoreWithoutCast`.
https://github.com/pytorch/pytorch/blob/bce4c82f0d2e4d36b96e5e10314c3bae16ed705e/aten/src/ATen/native/cuda/MemoryAccess.cuh#L263
Integral inputs cannot be promoted to floating types for CUDA inputs, and this check may be removed once the `gpu_kernel_multiple_outputs` is enhanced with casting support.